libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py

```diff
@@ -123,94 +401,128 @@ def _call(
-            header.pop("content_type", DEFAULT_CONTENT_TYPE_JSON)
-            or DEFAULT_CONTENT_TYPE_JSON
-        )
-        request_kwargs = {"json": data}
-        request_kwargs["headers"] = header
-        timeout = kwargs.pop("timeout", DEFAULT_TIME_OUT)
-
-        attempts = 0
-        while attempts < 2:
-            request_kwargs["auth"] = self.auth.get("signer")
-            response = requests.post(
-                endpoint, timeout=timeout, **request_kwargs, **kwargs
-            )
-            if response.status_code == 401:
-                self._refresh_signer()
-                attempts += 1
-                continue
-            break
+            An iterator of GenerationChunks.
 
-        try:
-            response.raise_for_status()
-            response_json = response.json()
 
-        except Exception:
-            logger.error(
-                "DEBUG INFO: request_kwargs=%s, status_code=%s, content=%s",
-                request_kwargs,
-                response.status_code,
-                response.content,
-            )
-            raise
+        Example:
 
-        return response_json
+            .. code-block:: python
+
+            async for chunk in oci_md.astream(("Tell me a joke."):
+                print(chunk, end="", flush=True)
+
+        """
+        requests_kwargs = kwargs.pop("requests_kwargs", {})
+        self.streaming = True
+        params = self._invocation_params(stop, **kwargs)
+        body = self._construct_json_body(prompt, params)
+
+        async for line in await self.acompletion_with_retry(
+            data=body, run_manager=run_manager, stream=True, **requests_kwargs
+        ):
+            text = self._handle_sse_line(line)
+            if text:
+                chunk = GenerationChunk(text=text)
+                if run_manager:
+                    run_manager.on_llm_new_token(chunk.text, chunk=chunk)
+                yield chunk
 
     def _refresh_signer(self) -> None:
         if self.auth.get("signer", None) and hasattr(
             self.auth["signer"], "refresh_security_token"
         ):
             self.auth["signer"].refresh_security_token()
 
+    def _parse_stream(self, lines: Iterator[bytes]) -> Iterator[str]:
+        for line in lines:
+            _line = self._parse_stream_line(line)
+            if _line is not None:
+                yield _line
+
+    async def _parse_stream_async(
+        self,
+        lines: aiohttp.StreamReader,
+    ) -> AsyncIterator[str]:
+        async for line in lines:
+            _line = self._parse_stream_line(line)
+            if _line is not None:
+                yield _line
+
+    def _parse_stream_line(self, line: bytes) -> Optional[str]:
+        line = line.strip()
+        if line:
+            line = line.decode("utf-8")
+            if "[DONE]" in line:
+                return None
+
+            if line.startswith("data:"):
+                return line[6:] if line.startswith("data: ") else line[5:]
```

Comment:
I think it would be more safe to delete the data: and all the whitespaces after it in the line.

Reply:
If there is no ```"data:"``` in the line, shouldn't we just return ```line``` as is?

Reply:
Hi Dmitrii, I am pretty sure I encountered the case that start with `data:` without space, probably when the llamacpp still working. I can try with another deployment to replicate the case. 

When there is no "data:" in the line,  it hit the end of streaming. Yes, we can return line. I am returning None to avoid error from `json.load()`. Let me try again on this case to see if `json.load` will throw error.

Reply:
Fixed!